src/openads-appnexus/AppNexusConnector.js

```diff
@@ -6,103 +6,208 @@ import {
-      .then(astClient => astClient.loadTags())
-      .then(() => this._adRepository.find({id: domElementId}))
+      .then(() => this._astClient.showTag({targetId: id}))
+      .then(null)
   }
 
-  refresh({domElementId, placement, sizes, segmentation, native}) {
+  refresh({id, specification}) {
+    this._logger.debug(
+      this._logger.name,
+      '| refresh | id:',
+      id,
+      '| specification:',
+      specification
+    )
     return Promise.resolve()
-      .then(() => this._adRepository.remove({id: domElementId}))
-      .then(() => {
-        let updateData = (placement || sizes || segmentation || native) && {}
-        if (updateData) {
-          if (placement) updateData.invCode = placement
-          if (sizes) updateData.sizes = sizes
-          if (segmentation) updateData.keywords = segmentation
-          if (native) updateData.native = native
+      .then(() => this._adRepository.remove({id: id}))
+      .then(() => this._refreshDebouncer.debounce({input: {id, specification}}))
+      .then(() => this._adRepository.find({id: id}))
+  }
+
+  /**
+   * @param {object} inputArray
+   * @returns {Promise<{appnexusInputsArray: Array, prebidUnitsArray: Array} | never | void>}
+   * @private
+   */
+  _onLoadAdDebounce(loadAdInputs) {
+    this._logger.debug(
+      this._logger.name,
+      '| _onLoadAdDebounce | loadAdInputs:',
+      loadAdInputs
+    )
+    return Promise.resolve(loadAdInputs)
+      .then(this._buildNormalizedInputs)
+      .then(normalizedInputs => {
+        this._logger.debug(
+          this._logger.name,
+          '| _onLoadAdDebounce | normalizedInputs:',
+          normalizedInputs
+        )
+        normalizedInputs.tags.forEach(input =>
+          this._defineAppNexusTag({id: input.id, tag: input.data})
+        )
+        if (normalizedInputs.adUnits.length > 0) {
+          this._prebidClient.addAdUnits({adUnits: normalizedInputs.adUnits})
+          this._prebidClient.requestBids({
+            requestObj: {
+              timeout: TIMEOUT_PREBID,
+              bidsBackHandler: () => {
+                this._prebidClient.setTargetingForAst()
+                this._astClient.loadTags()
+              }
+            }
+          })
+        } else {
+          this._astClient.loadTags()
+        }
+      })
+      .catch(error => console.log(error))
+  }
+
+  _onRefreshDebounce(refreshInputs) {
+    this._logger.debug(
+      this._logger.name,
+      '| _onRefreshDebounce | refreshInputs:',
+      refreshInputs
+    )
+    return Promise.resolve(refreshInputs)
+      .then(this._buildNormalizedInputs)
+      .then(normalizedInputs => {
+        normalizedInputs.tags.forEach(tag =>
           this._astClient.modifyTag({
-            targetId: domElementId,
-            data: updateData
+            targetId: tag.id,
+            data: tag.data
+          })
+        )
+        if (normalizedInputs.adUnits.length > 0) {
+          this._prebidClient.requestBids({
+            requestObj: {
```

Comment:
the requestObj is not a named param in the prebidjs API reference, there's no sense to have it named here, it'll be better to send the requestObj as a whole parameters object, or parametrize the prebid client method with the 4 expected parameters, but not using requestObj as parameter :)

Reply:
@alextremp it is done. ^_^